### ISSUE TITLE
[Static] Target "ociotest" links to target "sampleicc::sampleicc" and  "utils::strings",  but the target was  not found. 

### ISSUE THREAD

**Issue by PhoebeHui** [AcademySoftwareFoundation/OpenColorIO#1581]
Hi Guys,

Related to https://github.com/microsoft/vcpkg/issues/21642, the user encountered the following failure when use opencoloio with static build, we noticed that target sampleicc::sampleicc and utils::strings exported in static build, but these looks private which can't be used and exported externally. we could repro this issue on windows. Could you please help take a look?

OpenColorIOTargets.cmake:
```
set_target_properties(OpenColorIO::OpenColorIO PROPERTIES
  INTERFACE_COMPILE_FEATURES "cxx_std_11"
  INTERFACE_LINK_LIBRARIES "OpenColorIO::OpenColorIOHeaders;\$<LINK_ONLY:expat::expat>;\$<LINK_ONLY:Imath::Imath>;\$<LINK_ONLY:pystring::pystring>;\$<LINK_ONLY:sampleicc::sampleicc>;\$<LINK_ONLY:utils::strings>;\$<LINK_ONLY:yaml-cpp>"
)
```

Failures:
```
CMake Error at /Users/matt/vcpkg/scripts/buildsystems/vcpkg.cmake:540 (_add_executable):
  Target "ociotest" links to target "sampleicc::sampleicc" but the target was
  not found.  Perhaps a find_package() call is missing for an IMPORTED
  target, or an ALIAS target is missing?
Call Stack (most recent call first):
  CMakeLists.txt:11 (add_executable)

CMake Error at /Users/matt/vcpkg/scripts/buildsystems/vcpkg.cmake:540 (_add_executable):
  Target "ociotest" links to target "utils::strings" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?
Call Stack (most recent call first):
  CMakeLists.txt:11 (add_executable)
```

Repro steps:
1. git clone https://github.com/PhoebeHui/vcpkg
2. cd vcpkg
3. git checkout dev/Phoebe/opencolorio
4. ./bootstrap-vcpkg.bat
5. ./vcpkg install opencolorio:x64-windows-static
6. ./vcpkg integrate install, copy the '-DCMAKE_TOOLCHAIN_FILE=xxxx', it will be used in test steps.
7. Create a test project that will link with OpenColorIO：
test\CMakeLists.txt
    ```
    cmake_minimum_required(VERSION 3.22)
    project(ocio-test LANGUAGES CXX)

    set(CMAKE_CXX_STANDARD 14)
    set(CMAKE_CXX_STANDARD_REQUIRED ON)
    set(CMAKE_CXX_EXTENSIONS OFF)

    find_package(OpenColorIO REQUIRED)
    add_executable(ociotest    main.cpp)

    target_link_libraries(ociotest PRIVATE OpenColorIO::OpenColorIO OpenColorIO::OpenColorIOHeaders)
    ```
test\main.cpp
    ```
    #include <OpenColorIO/OpenColorIO.h>

    int main()
    {
      OCIO_NAMESPACE::Config::CreateFromEnv();
      return 0;
    }
    ```
8.  Create test\build folder,  replace the '-DCMAKE_TOOLCHAIN_FILE=E:/vcpkg/PhoebeHui/vcpkg/scripts/buildsystems/vcpkg.cmake' with your actual path. Then execute the command:
```
cmake.exe  -DCMAKE_TOOLCHAIN_FILE=E:/vcpkg/PhoebeHui/vcpkg/scripts/buildsystems/vcpkg.cmake  -DVCPKG_TARGET_TRIPLET=x64-windows-static -DVCPKG_BUILD_TYPE=debug -G "Visual Studio 16 2019" ..
```

**Comment by hodoulp** [comment]
Tested on macOS 10.15.7 with cmake 3.20.1 (without `vcpkg`) from `main`, `RB-2.1` (i.e. version 2.1.1) and `RB-2.0` (i.e. version 2.0.3) branches, the `OpenColorIOTargets.cmake` file does not include private libraries.

From `RB-2.1`, for `OpenColorIOTargets.cmake`:
```
add_library(OpenColorIO::OpenColorIO SHARED IMPORTED)

set_target_properties(OpenColorIO::OpenColorIO PROPERTIES
  INTERFACE_COMPILE_FEATURES "cxx_std_11"
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
)
```
and for `OpenColorIOTargets-release.cmake`:
```
set_property(TARGET OpenColorIO::OpenColorIO APPEND PROPERTY IMPORTED_CONFIGURATIONS RELEASE)
set_target_properties(OpenColorIO::OpenColorIO PROPERTIES
  IMPORTED_LOCATION_RELEASE "${_IMPORT_PREFIX}/lib/libOpenColorIO.2.1.2.dylib"
  IMPORTED_SONAME_RELEASE "@rpath/libOpenColorIO.2.1.dylib"
  )
```

From `RB-2.0`, for `OpenColorIOTargets.cmake`:
```
add_library(OpenColorIO::OpenColorIO SHARED IMPORTED)

set_target_properties(OpenColorIO::OpenColorIO PROPERTIES
  INTERFACE_LINK_LIBRARIES "OpenColorIO::OpenColorIOHeaders"
)

# Create imported target OpenColorIO::OpenColorIOHeaders
add_library(OpenColorIO::OpenColorIOHeaders INTERFACE IMPORTED)

set_target_properties(OpenColorIO::OpenColorIOHeaders PROPERTIES
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
)
```
and for `OpenColorIOTargets-release.cmake`:
```
set_property(TARGET OpenColorIO::OpenColorIO APPEND PROPERTY IMPORTED_CONFIGURATIONS RELEASE)
set_target_properties(OpenColorIO::OpenColorIO PROPERTIES
  IMPORTED_LOCATION_RELEASE "${_IMPORT_PREFIX}/lib/libOpenColorIO.2.0.4.dylib"
  IMPORTED_SONAME_RELEASE "@rpath/libOpenColorIO.2.0.dylib"
  )
```

In conclusion I found no problem when compiling/installing from macOS using the latest `OpenColorIO` versions.

**Comment by PhoebeHui** [comment]
@hodoulp, thanks for your repsonse! Building OpenColorIO with shared doesn't has this issue, the static build has this problem.

OpenColorIOTargets.cmake
```
# Create imported target OpenColorIO::OpenColorIO
add_library(OpenColorIO::OpenColorIO STATIC IMPORTED)

set_target_properties(OpenColorIO::OpenColorIO PROPERTIES
  INTERFACE_COMPILE_FEATURES "cxx_std_11"
  INTERFACE_LINK_LIBRARIES "OpenColorIO::OpenColorIOHeaders;\$<LINK_ONLY:expat::expat>;\$<LINK_ONLY:Imath::Imath>;\$<LINK_ONLY:pystring::pystring>;\$<LINK_ONLY:sampleicc::sampleicc>;\$<LINK_ONLY:utils::strings>;\$<LINK_ONLY:yaml-cpp>"
)
```

**Comment by remia** [comment]
Thanks for the report @PhoebeHui, I can reproduce using static build under macOS. It seems that we are using private header-only libraries and linking them privately to the OpenColorIO library target but they still make their way to the public exported dependencies.

I found the following [CMake issue](https://gitlab.kitware.com/cmake/cmake/-/issues/15415) discussing about this and tried the fix at the end:

```
target_link_libraries(OpenColorIO
         expat::expat
         ${OCIO_HALF_LIB}
         pystring::pystring
-        sampleicc::sampleicc
-        utils::from_chars
-        utils::strings
+        "$<BUILD_INTERFACE:sampleicc::sampleicc>"
+        "$<BUILD_INTERFACE:utils::from_chars>"
+        "$<BUILD_INTERFACE:utils::strings>"
         yaml-cpp
 )
```

This generate the following:

```
set_target_properties(OpenColorIO::OpenColorIO PROPERTIES
  INTERFACE_COMPILE_FEATURES "cxx_std_11"
  INTERFACE_INCLUDE_DIRECTORIES "${_IMPORT_PREFIX}/include"
  INTERFACE_LINK_LIBRARIES "\$<LINK_ONLY:expat::expat>;\$<LINK_ONLY:Imath::Imath>;\$<LINK_ONLY:pystring::pystring>;\$<LINK_ONLY:>;\$<LINK_ONLY:>;\$<LINK_ONLY:>;\$<LINK_ONLY:yaml-cpp>;-framework Carbon;-framework IOKit"
)
```

This removes the private library but leave holes instead, I'm not sure why or if this could be an issue for client projects. What do you think, would you be able to test that?

PS: I was wondering why you updated the Config.cmake.in file in your [patch](https://github.com/PhoebeHui/vcpkg/commit/2303224908103002d94d23dc7c85297495dbdf06#diff-542fdc52559edbbc2696ad7511034917be93126596f1b383bf20dc7ea5ae45d1R50-R59), is that something requiring a fix upstream?

**Comment by PhoebeHui** [comment]
@remia, thanks for the patch!

It works, however, after the issue solved, I found another failure:
```
unresolved external symbol "__declspec(dllimport) public: static class std::shared_ptr<class OpenColorIO_v2_1::Config const > __cdecl OpenColorIO_v2_1::Config::CreateFromEnv(void)" (__imp_?CreateFromEnv@Config@Op
enColorIO_v2_1@@SA?AV?$shared_ptr@$$CBVConfig@OpenColorIO_v2_1@@@std@@XZ) referenced in function main
```

After I check the CMakeList.txt, it looks OpenColorIO_SKIP_IMPORTS should be exported, so it should be PUBLIC. It solves the problem. I think OpenColorIO_EXPORTS should be exported as well.

[line#https://github.com/AcademySoftwareFoundation/OpenColorIO/blob/main/src/OpenColorIO/CMakeLists.txt#L244]
(https://github.com/AcademySoftwareFoundation/OpenColorIO/blob/main/src/OpenColorIO/CMakeLists.txt#L243-L252)


>PS: I was wondering why you updated the Config.cmake.in file in your patch, is that something requiring a fix upstream?

The patch in Config.cmake.in required, since when user call OpenColorIO::OpenColorIO, the dependency libraries can't be found, opencolorio should handle its dependency.
